File: solutions/python/square-root/1/square_root.py
def square_root(number):
    # Use integer binary search to find the exact square root
    low, high = 1, number

    while low <= high:
        mid = (low + high) // 2
        sq = mid * mid

        if sq == number:
            return mid
        elif sq < number:
            low = mid + 1
        else:
            high = mid - 1

File: solutions/java/square-root/1/src/main/java/SquareRoot.java
class SquareRoot {

    int squareRoot(int number) {
        int left = 1;
        int right = number;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            long square = (long) mid * mid;

            if (square == number) {
                return mid;
            } else if (square < number) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1; // This line is never reached for valid Exercism inputs.
    }
}